File: leetcode/algorithm/binary-tree-zigzag-level-order-traversal.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

import time

class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if root is None :
            return []

        stack = [root]
        ans = []
        flag = 1
        while stack:
            temp = []
            print(stack)
            for _ in range(len(stack)):
                node = stack.pop(0)
                temp.append(node.val)
                if node.left:
                    stack.append(node.left)
                if node.right:
                    stack.append(node.right)
            if flag :
                ans.append(temp[::])
            else:
                ans.append(temp[::-1])
            flag = 1 - flag
        
        return ans

File: leetcode/algorithm/binary-tree-zigzag-level-order-traversal.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func zigzagLevelOrder(root *TreeNode) [][]int {
	queue := []*TreeNode{root}
	flag := true
	ans := [][]int{}
	if root == nil {
		return ans
	}

	for len(queue) != 0 {
		newqueue := []*TreeNode{}
		tmpAns := []int{}
		if flag {
			for i := 0; i < len(queue); i++ {
				tmpAns = append(tmpAns, queue[i].Val)
			}
		} else {
			for i := len(queue) - 1; i >= 0; i-- {
				tmpAns = append(tmpAns, queue[i].Val)
			}
		}

		for i := 0; i < len(queue); i++ {
			n := queue[i]
			if n.Left != nil {
				newqueue = append(newqueue, queue[i].Left)
			}
			if n.Right != nil {
				newqueue = append(newqueue, queue[i].Right)
			}
		}
		ans = append(ans, tmpAns)
		queue = newqueue
		flag = !flag
	}
	return ans
}
